### ISSUE TITLE
Discussion

### ISSUE THREAD

**Issue by antony-jr** [antony-jr/AppImageUpdaterBridge#1]
What will this library do ?
* Implement a simple to use interface to **AppImageUpdater** with Qt5 **signals** and **slots**
* Create a easier to use **C++ API**.
* Document the new **High Level API** , so everyone can use it
* Create an Async model.

The **goal** is to make the library **Qt** friendly ( I'm bad at naming! ).

Any Thoughts?

**Comment by probonopd** [comment]
👍  Maybe call this AppImageUpdateBridge to make clear it is about _updating_?

**Comment by probonopd** [comment]
(Moved from https://github.com/antony-jr/GHReleaseBridge/issues/1)

Hey @antony-jr please check out [libappimageupdate](https://github.com/AppImage/AppImageUpdate). It implements the update information mechanisms described in the [AppImageSpec](https://github.com/AppImage/AppImageSpec/blob/master/draft.md#update-information).

If you integrate [libappimageupdate](https://github.com/AppImage/AppImageUpdate/wiki/Guide-to-libappimageupdate) we might be interested. I am sure that @TheAssassin is happy to answer any questions you might have.

**Comment by antony-jr** [comment]
**There is a saying I like:**
> Perfection is Achieved Not When There Is Nothing More to Add, But When There Is Nothing Left to   
> Take Away   
> -- Antoine de Saint-Exupery
## We don't need the Magic Codes
As the updater is inbuilt inside the **AppImage** I think having it check the Magic codes is just a waste of time ( **please let me know if I am wrong?** ) ( for integrity we can use **SHA256** sums ).

**Reasons:** 
* The configuration can be done on a more reliable source like a json file inside the **AppImage** , 
this avoids the nasty code and its much faster and userfriendly.
*  The configuration can also be directly inputed in the code block. So this also proves integrity.
* We don't have to relay on **system specifics** , by avoiding the magic code search, I think we can
make the code elegant 

Any thoughts? @TheAssassin 

And thanks for the **zsync2** @TheAssassin , its really a life saver!

**Comment by probonopd** [comment]
>There is a saying I like:
>
>Perfection is Achieved Not When There Is Nothing More to Add, But When There Is Nothing Left to
Take Away
-- Antoine de Saint-Exupery

That could go on the AppImage homepage... 👍 

True, if AppImageUpdaterBridge is invoked from within the AppImage itself, then of course it doesn't need to check the AppImage magic bytes.

**Comment by TheAssassin** [comment]
> As the updater is inbuilt inside the AppImage I think having it check the Magic codes is just a waste of time ( please let me know if I am wrong? ) ( for integrity we can use SHA256 sums )

libappimageupdate does check for the magic bytes to determine the type which defines where it has to look for the update information. I don't think that this really tiny check is a "waste of time" or anything else, and it is actually necessary because it is the only way to gather this information (not only reliably, but at all, anything else would mean to perform checks for all types for _every_ file, but also won't scale well for future AppImage types).

Using JSON files (or anything similar to that) from *within* the AppImage doesn't make sense to me as we have a specification defining an update mechanism, that is supported by this library. Extracting files from an AppImage is a certainly more expensive operation than reading 2-3 bytes from a file's beginning. And as libappimageupdate is primarily built to update external AppImages (e.g., as a self contained tool (appimageupdatetool, AppImageUpdate), or some app store like applications), it makes sense to have this as default.

> for integrity we can use SHA256 sums

Magic numbers/bytes != integrity check. They are used to identify the AppImage type, but that's not related to any specific patching or updating code, only for determining meta information required to start the update processes. I don't know what you're trying to express with this sentence. Feel free to elaborate to clarify this point.

> The configuration can be done on a more reliable source like a json file inside the AppImage ,
this avoids the nasty code and its much faster and userfriendly.

If you want to change this update mechanism to be defined by a file inside the AppImage, that'd only work for self-updating AppImages, but your code might even be useful to update external AppImages. I don't think adding such a limitation is necessary.

Also, re user friendliness, when creating AppImages with appimagetool, you just have to pass the `-u` flag and give it the update information string, that process is well documented, and there's even a GitHub releases type which interprets what the GitHub releases API returns. That being built in makes the library  pretty portable and self contained.

libappimageupdate is built to serve the purpose of making use of the update processes specified in the AppImage specification. There won't be support for anything other than this in the library itself. If you really *insist* on your other ideas (which in my opinion aren't any superior (easier to use, more efficient, ...) to what AppImage provides), you'll have to write these bits yourself. Feel free to take a look at libappimageupdate's code and how it makes use of libzsync2.

> nasty code

Please elaborate on what you consider *nasty code* in relation to libappimageupdate, AppImageUpdate, the AppImage update mechanisms or anything related. I can't really reenact this part.

Oh, and by the way, thanks for the kind words about zsync2. I know it lacks most docs at the moment, but I read from your words that you figured out how to use it. Feel free to add some docs or write some requests for such docs in issues on GitHub.

**Comment by antony-jr** [comment]
@TheAssassin not extracting the appimage , if the program is executed ( i.e the Updater or the main program [ Both are same ]) , it is loaded to the memory , now the **AppImage** and our instance of the **App** i.e the **Updater** is detached and thus we require no operation calls to read the configuration and so this is much faster than reading and parsing (and the parsing is really not a good practice )  the **first 4 bytes** from a file in the disk.

**Comment by antony-jr** [comment]
we can also use **zsync's** sha1 checks on the final result , if wanted!

**Comment by antony-jr** [comment]
I think I should rewrite **zsync** for **Qt** with a clean API interface. So please wait!

**Comment by probonopd** [comment]
I thought that was just what @TheAssassin had done with zsync2?

**Comment by TheAssassin** [comment]
@probonopd zsync2 has a C++11 interface, so, standard types instead of the Qt versions (but actually it's all compatible...). I don't see the issue leading to the decision of @antony-jr to rewrite the original zsync (which has had a *lot* of issues, which took many weeks of intensive work to fix all those issues in zsync2. You shouldn't underestimate the amount of engineering that was necessary to create both libzsync2 and libappimageupdate.

**Comment by antony-jr** [comment]
Yeah , I'm fine with it , once I finish with **zsync** I can easily build the Updater with elegance.
I'm aiming for a api like this.

```
AIUpdaterBridge Bridge({
       {"username", "antony-jr"},
       {"repo" , "QArchive"},
       {"version", "0.0.1"},
       {"assetFormat", "QArchive-{version}.tar.gz"},
       {"installationPath", "./"},
       {"debug", false}
});
```
just simple as **GHReleaseBridge**.

**Comment by probonopd** [comment]
@antony-jr of course it is your decision but I would prefer to just have one supported version that we use for AppImage, so that we need to fix issues in just one rather than multiple locations. `libappimageupdate` (which uses `libzsync2` internally) was specifically developed with being used in GUI frontends including Qt ones in mind.

**Comment by antony-jr** [comment]
Then why mix it with **FlTk** , **Qt** is just perfect by its own.

**Comment by probonopd** [comment]
> I'm aiming for a api like this.

What should that API do?

Please be aware that the AppImageSpec specifies how update information is handled inside AppImages. All of the work has already been done by @TheAssassin. Please don't re-invent the wheel, just focus on making a Qt GUI for libappimageupdate which does all the real work.

**Comment by antony-jr** [comment]
yes thats just a template

**Comment by antony-jr** [comment]
I do use some code from @TheAssassin 's work.

**Comment by antony-jr** [comment]
**I'm Just wrapping the client** with Qt , I will be using **zsync-curl** as its much cleaner! and curl is not that bad.

**Comment by probonopd** [comment]
Please note that we want all the AppImages be updated in the same way. We don't want some AppImages that do not use update information as specified in AppImageSpec but use some other way of updating, because this breaks all other existing AppImage-handling and updating tools out there.

There are already applications that can update _all_ the AppImages on a system. If you invent your own thing, that would break.

**Comment by probonopd** [comment]
```
{"assetFormat", "QArchive-{version}.tar.gz"}
```

Please explain what that means. AppImages are never supposed to be inside a .tar.gz.

**Comment by antony-jr** [comment]
its just a template , I'm following the Specs. Just concentrating on understanding zsync for now.

**Comment by probonopd** [comment]
```
       {"username", "antony-jr"},
       {"repo" , "QArchive"},
```

Please explain what that means. AppImages are supposed to be updated from the update information stored inside the AppImage. How this works is specified in the AppImageSpec, and `libappimageupdate` has already implemented it.

**Comment by probonopd** [comment]
> its just a template

Yes but my point is - there is no need for a "template" since the information you are looking for is _already_ in the AppImage and in the tools we have made for dealing with AppImages, such as `appimagetool`. Before you invest a lot of time to create new things, please invest some time to learn about what is already there.

>  I'm following the Specs

:+1:  This is the relevant part: https://github.com/AppImage/AppImageSpec/blob/master/draft.md#update-information

**Comment by antony-jr** [comment]
okay

**Comment by TheAssassin** [comment]
@antony-jr you're constantly mixing up projects. Are you referring to zsync, zsync-curl or zsync2? One never knows what you're talking about.

Short history lesson:

  - zsync: original project, written in very old style C, designed as CLI tool, so a lot of static globals etc., not at all usable as a library by other projects
  - zsync-curl: zsync, but using cURL as the backend to have support for HTTPS connections. Has had the same issues as zsync, as it was pretty much the same code base.
  - zsync2: Mostly a rewrite of the relevant code, especially main code and a bunch of other problematic parts, provides a library with a clean C++11 API, everything's contained in instances of `ZSyncClient`, etc. Meant to be used by external projects. Has been developed to be used by AppImageUpdate, libappimageupdate and the tools which want to make use of those.

**Comment by antony-jr** [comment]
I first saw **zsync2** but the libs are not good with Qt , so I got into **zsync-curl**

**Comment by probonopd** [comment]
Can you please describe what exactly you mean by "the zsync2 libs are not good with Qt" and why you think that zsync-curl is more suitable for Qt?

Also, why do you think you need to interact with zsync2 directly at all? It should all be wrapped by libappimageupdate which would be what you would be using for Qt.

**Comment by antony-jr** [comment]
I think I will go with **zsync2** ?

**Comment by antony-jr** [comment]
I just thought , why use **open source libs** when **Qt** got them all , **cpr** , **args** , and **gtest**.

**Comment by antony-jr** [comment]
if we can replace those libs with Qt ,  we can easily create a amazing async model for **zsync2**

**Comment by antony-jr** [comment]
but now its not the time right ?

**Comment by antony-jr** [comment]
okay getting back to work!

**Comment by antony-jr** [comment]
@TheAssassin I will be using **zsync2** for now , if possible I will help you later when I finish this project.

**Comment by probonopd** [comment]
> I just thought , why use open source libs when Qt got them all , cpr , args , and gtest.

The reason is that we need it for non-Qt projects, too. So we are aiming at a an architectural separation between a toolkit-independent libraries and Qt, Gtk+,... GUIs. This way, we don't have to maintain the core code in more than one place.

**Comment by TheAssassin** [comment]
>  why use open source libs when Qt got them all , cpr , args , and gtest.

That's a very weird way to look at free software...

zsync2 will most likely *never* ever be built around Qt. Qt is a way too big dependency for this tool. args is an amazing command line parser, which works perfectly fine, is very compact, and can be linked statically without any hassle. Same goes for gtest, but it's way less relevant in such considerations, as it's only used for unit testing (which we don't have too many of anyway at the moment).

For AppImageUpdate, there were thoughts about building a Qt interface. But noone's thinking of writing core itself (libappimageupdate) against Qt. And you shouldn't expect this to change. You seem to be pretty confident that a rewrite in Qt would be desired, which is not the case...

**Comment by antony-jr** [comment]
**Qt Framework** has a async model. Its about self-updating right. If Qt is used for the libappimageupdate then we can build elegant Qt GUI which will work on all linux distros and thus using Qt is wise because when we use Qt GUI libs we get all the advantages of Networking , Argument Parsing and other things together so why waste them ?

**Comment by antony-jr** [comment]
if we can use only one framework then maintaining the code is much easier! Qt docs are good too..

**Comment by antony-jr** [comment]
Qt and AppImages are the best combination.

**Comment by probonopd** [comment]
> if we can use only one framework

We cannot.

AppImage and AppImageUpdate are not just for Qt apps, but we want the updater experience to have a native look and feel in Qt apps. Similar for other application frameworks. So we need to use Qt for Qt applications and Gtk+ for Gtk+ applications and Electron for Electron applications. 

> If Qt is used for the libappimageupdate then we can build elegant Qt GUI

And what would Gtk+ applications and Electron applications use? They would certainly not add a Qt dependency to their code just for the updating. Or would you accept a Gtk+ dependency in your Qt applications?

**Comment by antony-jr** [comment]
ok got it!

**Comment by antony-jr** [comment]
I am trying to do something like this , https://github.com/AppImage/AppImageUpdate/issues/30 . with Qt Only

**Comment by antony-jr** [comment]
I already done it with an electron app but just a prototype with python and Qt. [exercism-gui-installer](https://github.com/antony-jr/exercism-installer)

**Comment by antony-jr** [comment]
@TheAssassin Do I need the submodules for **CPR** , because cmake is throwing an error. I'm forced to clone the **entire curl lib and mongoose project** , but the instruction is not given in the readme of **zsync2** , so I'm confused! Do I need to clone the submodules of **CPR** ?

**Comment by probonopd** [comment]
@antony-jr what are you trying to do?

Please follow the instructions given in https://github.com/AppImage/AppImageUpdate/wiki/Guide-to-libappimageupdate - at which step are you running into the difficulties you describe?

**Comment by antony-jr** [comment]
got it.

**Comment by TheAssassin** [comment]
You need the submodules, yes. They aren't used by cpr actually, but the default CMake configuration expects them. A workaround is planned but not yet ready to use.

**Comment by antony-jr** [comment]
From the discussion I've finally come to a conclusion.
## What does this bridge provide ?
* As I said , using **libappimageupdate** , this header-only ( or a lib? ) will provide a **light-wieght** bridge to **Qt** , Thus it makes it easier to make **native** GUI.
* Even if the **App** is **not a Qt** based program we can still use something like a **AppWrapper** , where the **AppWrapper** reads the configuration from a **json** file or **inside its code block**.
This **AppWrapper** will open a **QProcess** when there is **no Update** And starts a GUI Updater when there is an Update thus **Even Non-Qt Projects** can implement the autoupdate without hassle, **AppRun** will point to **AppWrapper** and inturn the **AppWrapper** will point to the **program executable** (this configuration can also be inside the json file ,  self contained inside **AppWrapper** using **Qt Resource System** ) , Then finally the user can simply throw the **AppWrapper** , **Configuration files(json)** and their **executables** inside the **AppImage** , When there is a new update , **AppWrapper** is used to update the **AppImage** and thus the **new** configuration is provided from the update **itself**.
* Since the new update **replaces** **AppWrappers** or **The AppImages** configuration or magic codes , we don't need to have a central **db**

This is just a bridge and will **not** provide a **GUI** , it just makes it **easier** to integrate with **Qt Apps** , The **AppWrapper** was my **mainstream**.

**Note** : **GHReleaseBridge** was able to **Update AppImages Automatically** without **libappimageupdate** but as @probonopd  said I am replacing or reinventing this bridge for **zsync** and **libappimageupdate**.

**Comment by probonopd** [comment]
> Even if the App is not a Qt based program

Then the application will not want to bunde AppImageUpdaterBridge, but use something else with a native UI (e.g., Gtk+, Electron) instead. So I would take this out of scope.

> we can still use something like a AppWrapper , where the AppWrapper reads the configuration from a json file or inside its code block

The "configuration" is already part of the AppImage. It is called the "update information" and libappimageupdate already handles it for you. No need to reinvent.

(Skipped the sentences talking about AppWrapper and json files, since I think these are not needed)

> This is just a bridge and will not provide a GUI

Now I am entirely confused. I thought the whole point of this exercise was to take libappimageupdate and build a Qt GUI around it that applications can embed.

User Story: "I as an application author want to make my AppImage self-updateable from the menu, "Check for newer version and update..."

Something like this:
![](https://user-images.githubusercontent.com/2480569/31623386-8c950a70-b29f-11e7-894f-a257b5cc8bd9.png)

Or do you have something different in mind?

**Comment by antony-jr** [comment]
**As the title said** its a **Bridge** , think of it as a **core** to the **main GUI**.
I will creating this bridge because.
* **libappimageupdater** does not go well with signals and slots. and thus this part has to be wrapped with signals and slots to make a good **Qt GUI**. 
* Using this bridge I will also create a **GUI** , Some user may want to use their **imaginations** so they can also use this bridge ( I will document this too ).
* Having a clean logic first is the key to a great **GUI App** and thus I think I will have to concentrate on the **Core** and then move on to the **GUI** part.

**Comment by TheAssassin** [comment]
Basically, I think it is fairly simple to write a Qt-ish wrapper for libappimageupdate, which I think is what you're trying to do. libappimageupdate works asynchronously by creating threads when necessary, that works quite automagically. I think this is what you're trying to do, but with every new content, you kind of change your announcements...

**Comment by antony-jr** [comment]
@TheAssassin True. Thats why this is just going to be **header-only** , easier on the users and developers. Just copy the header , refer the docs , design your gui and bind the controls.

**The Default design would what we will be creating with this bridge**.

**Note** : libappimageupdate does not go well with signals and slots eventhough its asynchronously.
what about progress , its not in async , you have to loop through which will block the main thread and will cause some unexpected results. Not only **progress** but some things like **error** handling can be done cleanly with **signals** and **slots**.

i even gone through the source but the progress is not async , its like a series of sync without blocking!

**Comment by antony-jr** [comment]
@TheAssassin The wrapper would sound like a simple task but **a problem** should be broken down into small pieces. 
I initially got **inspired** by the issues with **AppImages** on **AutoUpdate**.
Thats why I've created a lot of **header-only libs** to solve this problem,  not only to **AppImages** but to all **Open Source** apps. This project was meant to be called **AppWrapper**.
My plan was to first build all the **logics** first then create a **central** **GUI APP** for updating which supports **Qt  Apps** , **AppImages** and **Github deployed Apps** ( like electron and other hybrid apps ).  { All meant to be cross-platform except **AppImage** }

**Note** : [GHReleaseBridge](https://antony-jr.github.io/GHReleaseBridge) can update **AppImages** without delta compression.

> Just letting you know the other options and why I'm building a bridge like this.

**Comment by TheAssassin** [comment]
There have been plans to replace the custom async API with the futures API C++11 provides, however due to lack of time those plans haven't been realised yet. The current system is well tested and works fine. I will keep working on this. The feature should be easier to map to Qt.

Talking about progress updating, well, you seem to be mixing up a few things here. The update itself is running in a thread in the background. When requesting status updates, one has to lock this thread for a very, very short time in order to fetch the relevant variables (update state (running, finished, errored etc.), progress indication etc.). This is a technical necessity to make the calls thread safe, and not mess up with the class in the background. I don't see the issue at all. The only "faster" solution would be to have a data type for the state/progress variables that allows parallel reads/writes, which adds even more complexity with little benefit in real life.

Really, you're overthinking the whole "efficiency" aspect at the moment. I'd call that "premature optimization", there isn't even code which implements the API as far as I can see, or even better, neutral benchmarks which would prove your points. For example, you've been arguing about reading 3 bytes of a file being slower than reading data from memory, which is obviously true. However, in real life, those differences are negligible (nanoseconds vs. milliseconds) for applications, unless running huge numbers of such processes in parallel (thinking of 1000+ here).

I don't really see what you mean when talking about  "async" progress updates by the way. Are you thinking of running those updates in the background (i.e., in an extra thread) as well? You could implement something like that with ease if necessary. But I wouldn't really do that.
Something I came up with some time ago is a callback based state update feature, i.e., you can set callbacks which are called whenever there is a progress and/or state update. That's easy to use, should work a little faster than the existing system, and if done properly, you can just bind that to functions updating your GUI etc. (although those calls will block the update process until they terminate, unless called asynchronously as well, which needs some synchronization (i.e., locking) as well to avoid race conditions, making this basically the same approach as what we have, and is even potentially slower than the existing system depending on the UI toolkit).
The way zsync(2) works is that a chunk is being processed, then the progress is calculated. That variable is what needs to be fetched. This leaves you with the problem of potentially parallel reads/writes to the variable, and that must not happen, hence using a lock to protect the variable. Improvements welcome, as long as they make sense.

> delta compression

There is no delta "compression". The delta update mechanism is not for compressing anything, it just selectively downloads whatever chunks in a file have changed, reusing local data, making this a lot more efficient. You should try to be more precise about your wordings.

**Comment by antony-jr** [comment]
> although those calls will block the update process until they terminate, unless called asynchronously as well, which needs some synchronization (i.e., locking) as well to avoid race conditions, making this basically the same approach as what we have, and is even potentially slower than the existing system depending on the UI toolkit

@TheAssassin thats why we use **signals** and **slots** , its not the same as callbacks. Remember **Qt Project** is also **Open Source** ,its dual licensed with **GPL** and **LGPL Licensed** ,  its like a **Open source** framework to **rule them all**. You only have to **buy** Qt if you modify the source and **don't** release them.(i.e **Qt's** source code , you can have your source as commercial too.. )

**Comment by antony-jr** [comment]
**UPDATE**

* Implemented a efficient way to get **appimage** update information!

Hey @TheAssassin you can also use this very small static lib to get the update information for **Type 2 AppImages** without **Objdump**. Please see [AppImageUpdateInformation](https://github.com/antony-jr/AIUpdateInformation)

The lib helps us to do it clean.
```
AppImageUpdateInformation AppImage("appimaged.AppImage")
if(AppImage.hasError()){
   std::cerr << "Has Error!\n";
   return -1;
}else{
auto Info = AppImage.getUpdateInformation();
}
```

**Comment by antony-jr** [comment]
**NOTE** : some **AppImageToolkit** from the releases returns **wrong** entries , like filenames , to see this issue , clone this [repo](https://github.com/antony-jr/AIUpdateInformation) and make the **example** , it only depends on **C++ 11** supported **Compiler**

**Comment by antony-jr** [comment]
**UPDATE**
* Implemented checking for new updates 
* on new updates emits a singal ``` updatesAvailable( appImage , SHA1Remote)```
* on no new updates emits a signal ```noUpdatesAvailable( appImage , SHA1Remote)```
* Users can use the ``` noUpdatesAvailable``` signal to start their main window.
* Currently this lib only depends on **Qt5 Framework** and nothing else.
* **Does it clean with a very cool async model** , does not use **Objdump**.
* Automatically extracts information from **AppImages** , type 1 , type2 and user defined config inside code block ( must follow **AppImageSpec** )

![Imgur](https://i.imgur.com/Q4ePRpi.png)


**FUTURE DEVELOPMENTS**
* Resolve Url 
* Use zsync to fetch the remaining blocks
* Refactor Code.

**Note** : Since this lib is only needed for **Updates** , we don't even need the entire **zsync2** library , So ended up using the **legacy** version as I only need to recieve the file.

**Comment by antony-jr** [comment]
# News 
Hey @probonopd @TheAssassin Thank you for your support :heart: , The **API** is almost done. small **nitpicks** are there , So I should fix it.

**DOES NOT USE ZSYNC2 OR LIBAPPIMAGEUPDATE** , as they are so heavy.

**Only Depends on Qt5 , Curl  and C++ 11 Support** 

**Witness it with your own eyes!**
```
#include <QCoreApplication>
#include <AIUpdaterBridge.hpp>

int main(int argc , char **argv)
{
        // There is a progress signal too.
        QCoreApplication app(argc , argv);
        AIUpdaterBridge Bridge;
        QObject::connect(&Bridge , &AIUpdaterBridge::updatesAvailable,
        [&](QString appimage , QString sha1){
                Bridge.start();
        });
        Bridge.doDebug(true);
        Bridge.setAppImageUpdateInformation("appimaged-x86_64.AppImage");
        // or you can use this too , in-block versioning , sometimes useful
        /*
        Bridge.setAppImageUpdateInformation({
                     {"transport"  , "gh-releases-zsync"},
                     {"username" , "AppImage"},
                     {"repo"          , "AppImageToolKit"},
                      // Wildcard support!
                     {"filename"    , "AppImageTool-x86_64*.AppImage.zsync"},
                     {"tag"             , "continous" }
        }); // Thats it
        */
        /* or load from config.json
        QFile config("config.json"); config.open(QIODevice::ReadOnly);
        Bridge.setAppImageUpdateInformation(
                              QJsonDocument::fromJson(config.readAll()).object()
         ); // Thats it
         */
        return app.exec();
}
```

## What does this API Provide ?

* A simple to use interface to **AppImageUpdater** with Qt5 **signals** and **slots**
* A Documentation in Qt Style for convinience.
* Completely Async ( I even made a seperate thread for small loops , because even a nano secs can cause a lag in GUI )
* A Much robust error handling.
* Elegant Design.
* Very Very **Light-Wieght**.
* True Cross Platform ( i.e runs the same on **BSD** , **RedHat** , **Fedora** , **Ubuntu** and etc)

## What did I do ?

* Used **zsync-curl** for zsync support.
* Used **QAIUpdateInformation** for Update Information retrival. ( This is also made by me 👍 )
* Used **Qt5 Networking** and **Qt5 Core** for Networking and Other system specific things.

As the **zsync2** and **libappimageupdate** is not used , this is **100%** clean. Only Requires **Qt5**.
( **Note** : According to Qt , you can even sell software made by qt for **free** , but it will only cost when you change **Qt** 's source and not release it )

## What I have to do ?
* Do Documentation.
* Refactor the code.
* Add Support for **bintray** ( It would be helpful if I can get some example links )

**Comment by probonopd** [comment]
Hi @antony-jr can you please upload the version from the screenshot above so that we can test it? Thanks.

**Comment by probonopd** [comment]
> Only Requires Qt5.

Very interesting.

> or load from config.json

Why would one need or want this? AppImages should contain update information, so that this information always "travels" with the AppImage itself. We should not rely on external json files. To make a beautiful solution, "less is more". So I would remove the json thing. Especially since it will not be supported in other AppImage-related tools.

> bintray ( It would be helpful if I can get some example links )

Here you go:
https://bintray.com/probono/AppImages/Leafpad/0.8.18.1.glibc2.4#files (the links to the AppImage and its zsync file are inside there)

**Comment by antony-jr** [comment]
# To Test AppImageUpdaterBridge
**Depends On** : Qt5 , Curl and C++ 11 Supported Compiler. CMake ( Build Tool )

```
 $ git clone --recursive https://github.com/antony-jr/Test-AppImageUpdater
 $ cd Test-AppImageUpdater # Do not delete build directory
 $ cd build 
 $ cmake ..
 $ make -j4 # more if you have powerful pc.
 $ ./Updater # See the magic
```

**Comment by probonopd** [comment]
Having issues trying to compile this, please see https://travis-ci.org/probonopd/Test-AppImageUpdater/builds/325421803. Thanks.

**Comment by antony-jr** [comment]
sorry @probonopd I'm using ssh for **git pushes** so I assume that I need to create submodules without **ssh** , with normal **https** , Thus it failed to clone. Please wait.

**Comment by antony-jr** [comment]
Please Try this version!

# To Test AppImageUpdaterBridge
**Depends On** : Qt5 , Curl and C++ 11 Supported Compiler. CMake ( Build Tool )

```
 $ git clone https://github.com/antony-jr/Test-AppImageUpdater
 $ cd Test-AppImageUpdater # Do not delete build directory
 $ cd build 
 $ cmake ..
 $ make -j4 # more if you have powerful pc.
 $ ./Updater # See the magic
```

**Comment by antony-jr** [comment]
@probonopd this is how the test looks.

<p align="center">
<img src="https://imgur.com/download/lnMQRkP" width=auto height=auto />
</p>

**Comment by probonopd** [comment]
Interesting. Can you make it so that one can pass in a path to the AppImage to be updated in `$1`, e.g.,

```
./Updater /path/to/My.AppImage
```

Thanks.

**Comment by probonopd** [comment]
Looks like it doesn't truncate the update information before the first `0x00`:

```
# Copy Test-AppImageUpdater to the filename that it wants to update
me@host:~$ cp Downloads/Test-git.d36acda-x86_64.AppImage appimaged-x86_64.AppImage

# The run the updater
me@host:~$ Downloads/Test-git.d36acda-x86_64.AppImage QAIUpdateInformation:: Update information found.

QAIUpdateInformation::  "gh-releases-zsync|probonopd|Test-AppImageUpdater|continuous|Test-AppImageUpdater*-x86_64.AppImage.zsync\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"
AIUpdaterBridge:: github release link :: QUrl("https://api.github.com/repos/probonopd/Test-AppImageUpdater/releases/tags/continuous")
AIUpdaterBridge:: network error ::  QNetworkReply::NetworkError(ContentNotFoundError)
```

**Comment by antony-jr** [comment]
@probonopd I thought I fixed it  , its working fine for me.

**Comment by antony-jr** [comment]
>  Can you make it so that one can pass in a path to the AppImage to be updated in
sure 

I will do that , its simple too!

**Comment by probonopd** [comment]
Great. As we can see, the latest build was uploaded to
https://transfer.sh/Kuaby/Test-git.28cfe54-x86_64.AppImage

Source:
https://travis-ci.org/antony-jr/Test-AppImageUpdater/builds/325507604#L858

**Comment by antony-jr** [comment]
@probonopd buts its not working on pc but compiled version works
```
./Updater Test-git.28cfe54-x86_64.AppImage 
AppImage Version( "6b910c02f7b57471eac8fd87a632947ecb1a7fbc" )
Starting Application!
```


```
./Test-git.28cfe54-x86_64.AppImage Leafpad-0.8.18.1.glibc2.4-x86_64.AppImage 
./Test-git.28cfe54-x86_64.AppImage: relocation error: /usr/lib/libk5crypto.so.3: symbol   
 k5_os_mutex_destroy, version krb5support_0_MIT not defined in file libkrb5support.so.0   
with link time reference

```

So can this be considered self-updating appimage implementation!

**Comment by probonopd** [comment]
>  krb5support_0_MIT

This has to do with Kerberos.

Do we really need Kerberos? This is also something @TheAssassin already had solved.

**Comment by antony-jr** [comment]
**Hmm , I don't know I suppose its a qt dependency..** I will look into it!

**Comment by antony-jr** [comment]
Documentation is published , Please Refer the [documentation](https://antony-jr.github.io/AppImageUpdaterBridge) for more information on usage.

**Comment by probonopd** [comment]
Congrats @antony-jr for having released 2.0.0 of QAppImageUpdate with awesome new features!

Please consider sending a pull request for https://github.com/AppImage/awesome-appimage to represent it correctly there.

Keep up your great work!
